bridge/slack/helpers.go

```diff
@@ -38,27 +38,70 @@ func (b *Bslack) getChannel(channel string) (*slack.Channel, error) {
 }
 
 func (b *Bslack) getChannelByName(name string) (*slack.Channel, error) {
-	if b.channels == nil {
-		return nil, fmt.Errorf("%s: channel %s not found (no channels found)", b.Account, name)
-	}
-	for _, channel := range b.channels {
-		if channel.Name == name {
-			return &channel, nil
-		}
+	b.channelsMutex.RLock()
+	defer b.channelsMutex.RUnlock()
+
+	if channel, ok := b.channelsByName[name]; ok {
+		return channel, nil
 	}
 	return nil, fmt.Errorf("%s: channel %s not found", b.Account, name)
 }
 
 func (b *Bslack) getChannelByID(ID string) (*slack.Channel, error) {
-	if b.channels == nil {
-		return nil, fmt.Errorf("%s: channel %s not found (no channels found)", b.Account, ID)
+	b.channelsMutex.RLock()
+	defer b.channelsMutex.RUnlock()
+
+	if channel, ok := b.channelsByID[ID]; ok {
+		return channel, nil
+	}
+	return nil, fmt.Errorf("%s: channel %s not found", b.Account, ID)
+}
+
+func (b *Bslack) populateUsers() {
+	users, err := b.sc.GetUsers()
+	if err != nil {
+		b.Log.Errorf("Could not reload users: %#v", err)
+		return
+	}
+
+	newUsers := map[string]*slack.User{}
+	for _, user := range users {
+		newUsers[user.ID] = &user
 	}
-	for _, channel := range b.channels {
-		if channel.ID == ID {
-			return &channel, nil
+
+	b.usersMutex.Lock()
+	defer b.usersMutex.Unlock()
+	b.users = newUsers
+}
+
+func (b *Bslack) populateChannels() {
+	newChannelsByID := map[string]*slack.Channel{}
+	newChannelsByName := map[string]*slack.Channel{}
+
+	queryParams := &slack.GetConversationsParameters{
+		ExcludeArchived: "true",
+		Types:           []string{"public_channel,private_channel,mpim,im"},
+	}
+	for {
+		channels, nextCursor, err := b.sc.GetConversations(queryParams)
+		if err != nil {
+			b.Log.Errorf("Could not reload channels: %#v", err)
+			return
 		}
+		for _, channel := range channels {
+			newChannelsByID[channel.ID] = &channel
+			newChannelsByName[channel.Name] = &channel
```

Comment:
Seems that `IsIM:true` channels all have the same `""` name

Reply:
Ok. Will look into that tomorrow. 👍

Reply:
also channel gets reused in this loop (this is why my test bridge failed), it's always the same pointer
possible solution
``` 
for _, channel := range channels {
                lchannel := channel
		newChannelsByID[channel.ID] = &lchannel
		newChannelsByName[channel.Name] = &lchannel
}
```

Reply:
Had taken your suggestion but updated it to what I propose now:
```go
		for i := 0; i < len(channels); i++ {
			newChannelsByID[channels[i].ID] = &channels[i]
			newChannelsByName[channels[i].Name] = &channels[i]
		}
```

This has the advantage over the one proposed above of not having any copies, instead of two copies per channel that is iterated over (first copy into the `channel` variable, second into the iterations instance of `lchannel`).

Reply:
Yes, good thinking, but it comes with a loss of readability. (You could also do `for i := range channels { ... }` )
But I'm not going to nitpick over this, it's not like my code is that readable either :)
Looks good to merge now, thanks!

Reply:
Oops, not yet, the issue with the `IsIm:true` channels still exist